### PR TITLE
operator: use infrastructure.config.openshfit.io for platformtype

### PR DESCRIPTION
**Pull request by abhinavdahiya**
cluster operators must use the config.openshit.io as source for their top level config.

/cc @derekwaynecarr @enxebre

### CODE REVIEW

File: pkg/operator/config.go [hunk 1]
```diff
@@ -60,86 +49,21 @@ type Images struct {
...
-			nonNilFields = nonNilFields + 1
-		}
-		if nonNilFields > 1 {
-			return "", fmt.Errorf("more than one platform provider given")
-		}
-	}
-
-	if installConfig.AWS != nil {
+func getProviderFromInfrastructure(infra *configv1.Infrastructure) (Provider, error) {
+	switch infra.Status.Platform {
```
Comment (enxebre) [inline pkg/operator/config.go]: can't this function now just return `infra.Status.Platform` and we completely drop `type Provider string`?

File: config/kubemark-config-infra.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  - name: v1
+    served: true
+storage: true
+---
+apiVersion: config.openshift.io/v1
+kind: Infrastructure
+metadata:
+  name: cluster
+status:
+  platform: AWS
```
Comment (enxebre) [inline config/kubemark-config-infra.yaml]: this need to be 'kubemark' cc @ingvagabund

Comment (ingvagabund) [reply config/kubemark-config-infra.yaml]: Right, if not set properly our Jenkins CI will start to fail.

Comment (ingvagabund) [reply config/kubemark-config-infra.yaml]:
```
kubectl apply -f config/kubemark-config-infra.yaml
error: error validating "config/kubemark-config-infra.yaml": error validating data: [ValidationError(CustomResourceDefinition.spec.versions[0]): missing required field "storage" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.CustomResourceDefinitionVersion, ValidationError(CustomResourceDefinition): unknown field "storage" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.CustomResourceDefinition]; if you choose to ignore these errors, turn validation off with --validate=false
```

File: pkg/operator/config.go (hunk 1 above)
Comment (abhinavdahiya) [reply pkg/operator/config.go]: didn't want to change more than required. :)

File: config/kubemark-config-infra.yaml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  - name: v1
+    served: true
+  storage: true
+---
+apiVersion: config.openshift.io/v1
+kind: Infrastructure
+metadata:
+  name: cluster
+status:
+  platform: kubemark
```
Comment (enxebre) [inline config/kubemark-config-infra.yaml]: ```error: error validating "config/kubemark-config-infra.yaml": error validating data:[ValidationError(CustomResourceDefinition.spec): unknown field "storage" in io.k8s.apiextensions apiserver.pkg.apis.apiextensions.v1beta1.CustomResourceDefinitionSpec, ValidationError(CustomResourceDefinition.spec.versions[0]): missing required field "storage" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.CustomResourceDefinitionVersion]; if you choose to ignore these errors, turn validation off with --validate=false```
looks like storage field need to be set here @abhinavdahiya

Comment (enxebre) [reply config/kubemark-config-infra.yaml]: also I don't see status on the CRD def?

File: README.md [hunk 4]
```diff
@@ -161,22 +161,36 @@ INFO: For development and testing purposes only
...
+     - name: v1
+       served: true
+   storage: true
+   ---
+   apiVersion: config.openshift.io/v1
+   kind: Infrastructure
+   metadata:
+     name: cluster
+   status:
+     platform: AWS
```
Comment (ingvagabund) [inline README.md]: s/AWS/kubemark

File: pkg/operator/operator.go [hunk 5]
```diff
@@ -187,13 +188,13 @@ func (optr *Operator) sync(key string) error {
 	return optr.syncAll(*operatorConfig)
 }
 
-func (optr *Operator) maoConfigFromInstallConfig() (*OperatorConfig, error) {
-	installConfig, err := getInstallConfig(optr.kubeClient)
+func (optr *Operator) maoConfigFromInfrastructure() (*OperatorConfig, error) {
+	infra, err := optr.osClient.ConfigV1().Infrastructures().Get("cluster", metav1.GetOptions{})
```
Comment (enxebre) [inline pkg/operator/operator.go]: @abhinavdahiya e2e seems to be consistently failing, can't see anything meaningful in the logs. who/when creates the infra object?

Comment (enxebre) [reply pkg/operator/operator.go]: Managed to debug this by running installer locally from CI payload and running bastion to ssh into masters.
```
E0401 09:48:01.939450       1 operator.go:177] Failed getting operator config: infrastructures.config.openshift.io "cluster" is forbidden: User "system:serviceaccount:openshift-machine-api:default" cannot get resource "infrastructures" in API group "config.openshift.io" at the cluster scope
```
See if it goes green here https://github.com/openshift/machine-api-operator/pull/264/commits/df1d9c076b12eb42d36aa76b7f24a570a7938f19

**Comment by enxebre** [comment]
Closing in favour of https://github.com/openshift/machine-api-operator/pull/264

**Pull request closed without merging**